### DISCUSSION TITLE
cue injections

### DISCUSSION THREAD

**Discussion by romange** [cue-lang/cue#1176] [Q&A] [answered]
I am learning how to use injections for exporting the same `main.cue` file with different configurations.
It works if I export with `cue export prod.cue qa.cue main.cue -t prod`. 
Can I somehow import `prod.cue` and `qa.cue` from `main.cue`. Or maybe the import statement is intended for something else?  If I have to specify cue variable files explicitly then why do I need `-t prod` injection?  I could achieve the same by just adding a single variable file to export (i.e. qa.cue or prod.que)

```
// prod.cue
@if(prod)
env="prod"
....
----
// qa.cue
@if(qa)
env="qa"
.....
----
// main.cue
if env == "prod" {
}
```

**Comment by myitcv** [comment, marked as answer]
Hi @romange - thanks for raising this as a discussion; much easier for others to find this sort of question, and any answers.

Your question touches on an interesting point to which there are a few potential solutions. If I have understood correctly (and please let me know if I've got this wrong), you are looking to create variations of a configuration based on some parameters/environments. Tags are indeed one way of doing that, but they are not the only way, so it will be interesting to explore which is best for you.

One approach is to use different directories for the different variations:

```
# Eval the base package
exec cue eval .

# Eval prod
exec cue eval ./prod

# Eval qa
exec cue eval ./qa

-- cue.mod/module.cue --
module: "mod.com"
-- main.cue --
package main

import "path"

env:  string
host: string

// urlPath is agnostic of environment
urlPath: "/path/to/asset"

// url is derived from host and urlPath
url: "https://" + path.Join([host, urlPath])
-- prod/prod.cue --
package main

env:  "prod"
host: "prod.mod.com"
-- qa/qa.cue --
package main

env:  "qa"
host: "qa.prod.com"
```

This gives:

```
# Eval the base package (0.035s)
> exec cue eval .
[stdout]
import "path"

env:     string
host:    string
urlPath: "/path/to/asset"
url:     "https://" + path.Join([host, urlPath])
# Eval prod (0.040s)
> exec cue eval ./prod
[stdout]
env:     "prod"
host:    "prod.mod.com"
urlPath: "/path/to/asset"
url:     "https://prod.mod.com/path/to/asset"
# Eval qa (0.042s)
> exec cue eval ./qa
[stdout]
env:     "qa"
host:    "qa.prod.com"
urlPath: "/path/to/asset"
url:     "https://qa.prod.com/path/to/asset"
```

See the detail on [packages, modules and instances](https://cuelang.org/docs/concepts/packages/) for more info there, in particular this approach of the same package spanning multiple directories within a module.

Another approach is, as you suggest, using tags. Here is that same example rewritten to use tags:

```
# Eval the base package
exec cue eval .

# Eval prod
exec cue eval -t prod .

# Eval qa
exec cue eval -t qa .

-- cue.mod/module.cue --
module: "mod.com"
-- main.cue --
package main

import "path"

env:  string
host: string

// urlPath is agnostic of environment
urlPath: "/path/to/asset"

// url is derived from host and urlPath
url: "https://" + path.Join([host, urlPath])
-- prod.cue --
@if(prod)
package main

env:  "prod"
host: "prod.mod.com"
-- qa.cue --
@if(qa)
package main

env:  "qa"
host: "qa.prod.com"
```

Which gives:

```
# Eval the base package (0.041s)
> exec cue eval .
[stdout]
import "path"

env:     string
host:    string
urlPath: "/path/to/asset"
url:     "https://" + path.Join([host, urlPath])
# Eval prod (0.047s)
> exec cue eval -t prod .
[stdout]
env:     "prod"
host:    "prod.mod.com"
urlPath: "/path/to/asset"
url:     "https://prod.mod.com/path/to/asset"
# Eval qa (0.046s)
> exec cue eval -t qa .
[stdout]
env:     "qa"
host:    "qa.prod.com"
urlPath: "/path/to/asset"
url:     "https://qa.prod.com/path/to/asset"
```

As you can see, an`import` clause is not required because the files `main.cue`, `prod.cue` and `qa.cue` belong to the same package, and in this case are in the same directory. Which, as you can see, also means you don't need to specify the files themselves, just the path to the package you want to evaluate/export etc.

**Reply by romange** [reply]
Thank you for the extensive answer! 
I know my question was about using multiple files for configuration instantiation. 
Just to update, after I tried these options, I decided that for my specific purpose, having multiple config files  together with a module file and proper directory structure is a bit overkill. Luckily, cue supports simple tags as well, so I used them. see https://github.com/romange/docker/blob/master/packer/packer.cue#L29 and 
https://github.com/romange/docker/blob/master/packer/build.sh#L70 

Having said that, I will keep in mind how to use cue properly when configuration becomes more complicated.

**Reply by myitcv** [reply]
> Luckily, cue supports simple tags as well, so I used them

Great, yes I forgot that could really have been listed as an option above, although it's slightly different in style. 



> I will keep in mind how to use cue properly when configuration becomes more complicated

😄  - I don't think there is necessarily a "proper" or "right" way to do things per se. We're keen to learn about the different problems people are trying to solve, and see what patterns emerge in solutions. That might well, for example, involve rethinking the `cmd/cue` tooling and loading of CUE. So always keen to hear how/where CUE is getting in the way, could be doing more for you etc.
